### PR TITLE
Set Network MAC address for Lan7800.

### PR DESCRIPTION
**Pull request by henie**
In mailbox, queries and savesMAC address for LAN. When NDIS interface notification received, sets this MAC address in NetworkAddress in NDIS software key if the flag registry LanPropertyChange is not set.
During system boot, a service thread RpiLanPropertyChange is started and checks the flag RpiLanPropertyChange. If new MAC address is to be applied (RpiLanPropertyChange == 1), then restarts Lan7800, and sets flag RpiLanPropertyChange to 2.

Test results:
[Testcase.xlsx](https://github.com/ms-iot/rpi-iotcore/files/3016670/Testcase.xlsx)

### CODE REVIEW

File: drivers/mailbox/bcm2836/init.c [hunk 1]
```diff
@@ -102,6 +102,9 @@ Return Value:
     NTSTATUS value
 
 --*/
+
+WCHAR macAddrStrGlobal[13] = { 0 };
+
 _Use_decl_annotations_
 NTSTATUS RpiSetDeviceMacAddress (
```
Comment (dmcilvaney) [inline drivers/mailbox/bcm2836/init.c]: Is there a benefit to leaving the existing flow in? I don't know of any conflicts with this registry key, but if the new system is reliable I don't see a need to write the MAC twice.

File: drivers/RpiLanPropertyChange/bcm2836/service.cpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    {
+        // We currently do not have any workload that is affected by user logon/logoff
+        // consider: load swap database files
+    }
+    break;
+
+    case SERVICE_CONTROL_DEVICEEVENT:
+    {
+        // We currently do not have any workload that is affected by plug n play events
+        // consider: Oasis devices events (would allow a faster statup from callers perspective)
```
Comment (christopherco) [inline drivers/RpiLanPropertyChange/bcm2836/service.cpp]: Recommend to remove this comment or add something more relevant.

File: drivers/RpiLanPropertyChange/bcm2836/stdafx.cpp [hunk 3]
```diff
@@ -0,0 +1 @@
+#include "stdafx.h"
```
Comment (christopherco) [inline drivers/RpiLanPropertyChange/bcm2836/stdafx.cpp]: Add copyright header

File: drivers/RpiLanPropertyChange/bcm2836/stdafx.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// stdafx.h : include file for standard system include files,
```
Comment (christopherco) [inline drivers/RpiLanPropertyChange/bcm2836/stdafx.h]: Add copyright header

File: drivers/RpiLanPropertyChange/bcm2836/targetver.h [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#pragma once
```
Comment (christopherco) [inline drivers/RpiLanPropertyChange/bcm2836/targetver.h]: Add copyright header

File: drivers/mailbox/bcm2836/device.c [hunk 6]
```diff
@@ -430,6 +430,190 @@ Return Value:
...
+    OBJECT_ATTRIBUTES ObjectAttributes;  
+    HANDLE BaseKey = NULL;
+    HANDLE SubKey = NULL;
+    PKEY_VALUE_FULL_INFORMATION infoBuffer = NULL;
+    ULONG infoBufferLength = 0;
+    ULONG RequireBufferLength = 0;
+    DWORD value;
+
+    if(MacSet) {
+    return status;
```
Comment (christopherco) [inline drivers/mailbox/bcm2836/device.c]: Fix alignment

File: drivers/mailbox/bcm2836/device.c [hunk 7]
```diff
@@ -430,6 +430,190 @@ Return Value:
...
+    if (!NT_SUCCESS(status)){
+        BaseKey = NULL;
+        goto End;
+    }
+    
+    KeyName.Length = 8;
+    KeyName.MaximumLength = 8;
+    KeyName.Buffer = KeyInstance;
+
+    while(STATUS_OBJECT_NAME_NOT_FOUND != status)
```
Comment (christopherco) [inline drivers/mailbox/bcm2836/device.c]: Maybe add a comment with the high level overview of what the loop below achieves and how.

i.e. iterates through all enumerated ndis adapters looking for the LAN7800 instance and sets its MAC address if needed.

File: drivers/mailbox/bcm2836/init.c (hunk 1 above)
Comment (henie) [reply drivers/mailbox/bcm2836/init.c]: Thank you for your advice. As devices before 3B+ which is registry entry 0001 have different name, LAN9512/LAN9514 USB 2.0 to Ethernet 10/100 Adapter, I need to make a code change to add this device name into the driver and service.

File: drivers/RpiLanPropertyChange/bcm2836/RpiLanPropertyChange.vcxproj [hunk 8]
```diff
@@ -81,6 +83,13 @@
       <AdditionalDependencies>onecore.lib;kernel32.lib;user32.lib;%(AdditionalDependencies)</AdditionalDependencies>
       <SetChecksum>true</SetChecksum>
     </Link>
+    <PostBuildEvent>
+      <Command>set SIGN_OEM=1&amp;&amp;set&amp;&amp;"$(WDKContentRoot)\Tools\bin\i386\pkggen.exe" $(TargetName).pkg.xml /config:"$(WDKContentRoot)\Tools\bin\i386\pkggen.cfg.xml" /output:"$(OutDir)\$(TargetName) CBS\\" /version:1.0.0.0 /build:fre /cpu:ARM /variables:"SIGN_OEM=1;OEMNAME=Contoso;BSPARCH=ARM;_RELEASEDIR=$(OutDir);"</Command>
```
Comment (christopherco) [inline drivers/RpiLanPropertyChange/bcm2836/RpiLanPropertyChange.vcxproj]: @parameshbabu Is there a better way to invoke pkggen to generate a CAB for a service? PackageGen does not appear to be an option in the normal VS Studio properties like it is for drivers.

We are trying to include this service as part of the baseline BSP to fix a network MAC address issue.

Comment (henie) [reply drivers/RpiLanPropertyChange/bcm2836/RpiLanPropertyChange.vcxproj]: Yes, this is not good way. VS cannot generate packages for exe/dll like drivers. But if use pkggen, environment settings like test certs and signtools are required to setup.

**Pull request merged**